lib/ansible/module_utils/common/warnings.py

```diff
@@ -0,0 +1,25 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2019 Ansible Project
+# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+from ansible.module_utils.six import string_types
+
+global_warnings = []
+global_deprecations = []
```

Comment:
tldr; These need to be marked private.  Create a pair of simple accessor functions to give a read-only view of the data.  (Example implementation at the bottom)

These need to be marked private for three reasons: they're only implementation details of the framework, it's easy to imagine changes to the implementation (for instance, they could become a set.  Or you could start storing structured data like line numbers and calling functions inside of there), and we don't want people to modify them outside of the framework.

I see that you changed from private to public already as a result of a review comment.  The review comment does note something that's generally something to watch for but the solution of making it public is wrong.  Two things to note:

* In general, things outside of a framework touching private variables inside of the framework is wrong.  However, AnsibleModule is "special".  It is something that we are going to deprecate in the *long* term as we move things (via PRs like this) outside of it because AnsibleModule is an overgrown god object which violates the separation of concerns.  However, that combination of a god object which we are in the process of pulling apart does inately mean that it could be valid for it to touch private values inside of the pieces pulled from it.  For instance, if we decided that some of the existing API should only be supported in the deprecated AnsibleModule and implementing that API required using those private variables, then that would be a valid use case.

* However, in this case, we'll probably find that once we move the fail_json()/exit_json() style code outside of AnsibleModule, that code will need to still access the accumulated warnings and deprecations.  So what we want to do is provide some new API that can give non-legacy code access to the values stored in here without giving them the ability to modify it outside of the framework.  So the right implementation is probably accessor functions.  (Accessor is just a fancy way to say, I need a function to get the value of the data but it implies that the value cannot be modified unless there's a corresponding mutator function as well)

I think this would be a proper implementation of an accessor:
``` python
_global_warnings = []

def get_warning_messages():
    """Return a tuple of the warning messages accumulated over this run of the program"""
    return tuple(_global_warnings)
```

* The `get_` prefix to the function name is one thing that clues the user in that this is an accessor.
* Transforming the list into a tuple tells the user that they cannot modify the warnings via this function
* Using `_messages` in the function name allows you the leeway to write a new function that returns additional information should that be added in the future.

Reply:
So this is just a convention rather than a protection mechanism? I don't see anything that is preventing direct modification or use of the variables marked as private. How would someone know not to modify/reference `warnings._global_warnings` directly and instead go through the accessor/mutator functions?

Reply:
Right.  By naming with a leading underscore, you are marking it as private by convention but python doesn't build in anything to prevent someone from accessing it anyways.